lib/bakeAmbientOcclusion.js

```diff
@@ -0,0 +1,586 @@
+    randomDirection.y = t;
+    randomDirection.z = v * Math.sin(u);
+
+    // Orient with texel's normal in world space
+    var theta = Math.acos(normal.y); // dot product of normal with y-up is normal.y
+    var axis = Cartesian3.cross(randomDirection, normal, cartesian3Scratch1);
+    var rotation = Quaternion.fromAxisAngle(axis, theta, quaternionScratch);
+    var matrix = Matrix3.fromQuaternion(rotation, matrix3Scratch);
+
+    scratchRay.origin = position;
+    scratchRay.direction = Matrix3.multiplyByVector(matrix, randomDirection, scratchRay.direction);
+    return scratchRay;
+}
+
+// Borrowed straight from Cesium/Source/Core/IntersectionTests
+var scratchEdge0 = new Cartesian3();
+var scratchEdge1 = new Cartesian3();
+var scratchPVec = new Cartesian3();
+var scratchTVec = new Cartesian3();
+var scratchQVec = new Cartesian3();
+
+function rayTriangle(ray, p0, p1, p2, cullBackFaces) {
+    //>>includeStart('debug', pragmas.debug);
+    if (!defined(ray)) {
+        throw new DeveloperError('ray is required.');
+    }
+    if (!defined(p0)) {
+        throw new DeveloperError('p0 is required.');
+    }
+    if (!defined(p1)) {
+        throw new DeveloperError('p1 is required.');
+    }
+    if (!defined(p2)) {
+        throw new DeveloperError('p2 is required.');
+    }
+    //>>includeEnd('debug');
+
+    cullBackFaces = defaultValue(cullBackFaces, false);
+
+    var origin = ray.origin;
+    var direction = ray.direction;
+
+    var edge0 = Cartesian3.subtract(p1, p0, scratchEdge0);
+    var edge1 = Cartesian3.subtract(p2, p0, scratchEdge1);
+
+    var p = Cartesian3.cross(direction, edge1, scratchPVec);
+    var det = Cartesian3.dot(edge0, p);
+
+    var tvec;
+    var q;
+
+    var u;
+    var v;
+    var t;
+
+    if (cullBackFaces) {
+        if (det < CesiumMath.EPSILON6) {
+            return undefined;
+        }
+
+        tvec = Cartesian3.subtract(origin, p0, scratchTVec);
+        u = Cartesian3.dot(tvec, p);
+        if (u < 0.0 || u > det) {
+            return undefined;
+        }
+
+        q = Cartesian3.cross(tvec, edge0, scratchQVec);
+
+        v = Cartesian3.dot(direction, q);
+        if (v < 0.0 || u + v > det) {
+            return undefined;
+        }
+
+        t = Cartesian3.dot(edge1, q) / det;
+    } else {
+        if (Math.abs(det) < CesiumMath.EPSILON6) {
+            return undefined;
+        }
+        var invDet = 1.0 / det;
+
+        tvec = Cartesian3.subtract(origin, p0, scratchTVec);
+        u = Cartesian3.dot(tvec, p) * invDet;
+        if (u < 0.0 || u > 1.0) {
+            return undefined;
+        }
+
+        q = Cartesian3.cross(tvec, edge0, scratchQVec);
+
+        v = Cartesian3.dot(direction, q) * invDet;
+        if (v < 0.0 || u + v > 1.0) {
+            return undefined;
+        }
+
+        t = Cartesian3.dot(edge1, q) * invDet;
+    }
+
+    return t;
+}
+
+function sumBarycentric(barycentric, vector0, vector1, vector2, scratch, result) {
```

Comment:
Instead of passing in the scratch, just use a file-scoped scratch.  We generally do not pass scratch variables as parameters.